python-spec/src/somabase/data.py

```diff
@@ -0,0 +1,220 @@
+"""Definitions of data storage interfaces for SOMA implementations.
+
+SOMA users should ordinarily not need to import this module directly; relevant
+members will be exported to the ``somabase`` namespace.
+"""
+
+import abc
+from typing import Any, Iterable, Optional, Sequence, Tuple, Union
+
+import attrs
+import pyarrow
+from typing_extensions import Literal
+
+from somabase import base
+from somabase import options
+
+
+class DataFrame(base.SOMAObject, metaclass=abc.ABCMeta):
+    """A multi-column table with a user-defined schema."""
+
+    __slots__ = ()
+
+    # Data operations
+
+    @abc.abstractmethod
+    def read(
+        self,
+        ids,  # TODO: Specify type
+        column_names: Optional[Sequence[str]] = None,
+        *,
+        batch_size: options.BatchSize = options.BatchSize(),
+        partitions: Optional[options.ReadPartitions] = None,
+        result_order: Optional[options.ResultOrder] = None,
+        value_filter: Optional[str] = None,
+        platform_config: Optional[options.PlatformConfig] = None,
+    ) -> Iterable[pyarrow.Table]:
+        """Reads a user-defined slice of data into Arrow tables."""
+        raise NotImplementedError()
+
+    @abc.abstractmethod
+    def write(
+        self,
+        values: Union[pyarrow.RecordBatch, pyarrow.Table],
+        platform_config: Optional[options.PlatformConfig] = None,
+    ) -> None:
+        """Writes values to the data store."""
+        raise NotImplementedError()
+
+    # Metadata operations
+
+    @property
+    @abc.abstractmethod
+    def schema(self) -> pyarrow.Schema:
+        """The schema of the data in this dataframe."""
+        raise NotImplementedError()
+
+    @property
+    @abc.abstractmethod
+    def index_column_names(self) -> Tuple[str, ...]:
+        """The names of the index (dimension) columns."""
+        raise NotImplementedError()
+
+    # Basic operations
+
+    @property
+    def soma_type(self) -> Literal["SOMADataFrame"]:
+        return "SOMADataFrame"
+
+
+@attrs.define(frozen=True)
+class CoordsData:
+    """A slice of data returned from a SOMA dense read."""
+
+    coords: Any  # TODO: Define type.
+    """The coordinates of the slice that was read."""
+
+    data: pyarrow.Tensor
+    """The data that was read."""
+
+
+class NDArray(base.SOMAObject, metaclass=abc.ABCMeta):
+    """Common behaviors of N-dimensional arrays of a single primitive type."""
+
+    __slots__ = ()
+
+    # Data operations (listed alphabetically)
+
+    @abc.abstractmethod
+    def read_coo(
```

Comment:
Couple of issues here:
* soma.DenseNDArray read operation can't return an iterator, as it only returns a well-defined slice, whereas SparseNDArray _does_ return an iterator.
* ideally, the naming scheme would be normalized to allow for a variety of ecosystem/data model returned, as well as "iterator" or "no iterator"

`read_coo` doesn't tell me if this is an Arrow Tensor, Scipy COO spmatrix, COO data in a Table, etc.  It is probably fine if "Arrow" is the default ecosystem and implied, but I'd like to define the other typical cases.

Maybe something like:  <op>_all_"as"_<ecosystem>, where _as_ecosystem and _all are optional, and:
* Arrow is implied if ecosystem not stated stated
* For NDArray and DataFrame, Iterator is implied if no _all

So, examples for SparseNDArray:
* read_coo - iterator of Arrow Sparse Tensor
* read_table - iterator of Arrow Table
* read_dense - iterator of Arrow dense tensor
* read_as_scipy - iterator of scipy.sparse.coo_array (or coo_matrix)
* read_as_numpy - iterator of densified Numpy NdArray
* read_table_all - flatten output of read_table
* read_coo_all - flatten output of read_coo
etc.

DenseNDArray would be slightly different as it is currently not iterable, and would support:
* read_dense - Arrow Tensor (_not_ an iterator)
* read_dense_as_numpy - NDArray (_not_ an iterator)
etc.

Thoughts?

Reply:
This could end up being a huge number of methods to attach to each of these classes. Would it make sense to have something like:

```python
class SparseNDArray(...):
  # ...
  def read(self, ...) -> "SparseResult":
    ...


class SparseResult:
  # ...
  def as_coos(self) -> "ResultIterable[arrow.Tensor]":
    ...

  def as_tables(self) -> "ResultIterable[arrow.Table]":
    ...

class ResultIterable(Iterable[_T]):
  # Iterable methods for _T

  def flatten(self) -> _T:
    """Flattens all the results from this into one value."""
  # Alternately: .all
```

Then, using it would look something like this:

```python
my_table = some_nd_array.read(...).as_tables().flatten()
# or
for partial in some_nd_array.read(...).as_scipys():
  ...
```

These are just some first ideas for names, but I think doing something like this would be useful to separate the querying/filtering/loading results etc. part from the converting results to the desired format part (and make the method set of NDArray and friends much less intimidating).

I think this also means we could implement `ResultIterable` for each type as reusable wrappers around `Iterable`, since no matter what backend you’re, flattening a sequence of arrow `Table`s together ends up being the same thing:

```python

class Tables(ResultIterable[arrow.Table]):

  # __next__ and __iter__ forward to wrapped iterable

  def flatten(self) -> arrow.Table:
    # the actual code which actually flattens arrow Tables
    # which I am too lazy to look up what function it is
   ...
```

Reply:
I am definitely open to this idea!  At first glance, it appears cleaner from a user vantage (eg, the patterns learned with one type apply to others).  I'm not sure there are many places where we get a shared implementation, but that does not diminish the idea at all.

If we move to a API that supports explicit use of \_\_getitem__ (as proposed in single-cell-data/TileDB-SOMA#573), how would these be integrated together?  Or are they exclusive?